### ISSUE TITLE
`Page Components` could not catch a `server action` error.

### ISSUE THREAD

**Issue by b0jun** [vercel/next.js#63520]
### Link to the code that reproduces this issue

https://codesandbox.io/p/github/b0jun/next-server-action-error

### To Reproduce

1. The server action declared an error as follows:
```
/* src/actions/actions.ts */
'use server';
export async function getPost() {
	try {
		// API Call
		throw new Error('Email already exists', {
			cause: { status: 400, code: 'BA001' },
		});
	} catch (error) {
		throw error;
	}
}

```

2. The server action was called on the page as follows:
```
useEffect(() => {
		const callApi = async () => {
			try {
				await getPost();
			} catch (error: any) {
				console.error(error.cause); // undefined
				setErrorStatus(error.cause?.status); // expected value: 400
				setErrorCode(error.cause?.code); // expected value: 'BA001'
			}
		};

		callApi();
	}, []);
```

### Current vs. Expected behavior

In `error.cause`, the value I declared appears as `undefined`.
If I remove the `use server` statement, the state contains the value I want, but this is not the case in server actions.

Why are errors declared in server actions ignored?

In order to send an error to the page in object form from the server action, do I need to convert it to JSON and send it?
```
throw new Error(JSON.stringify({status: 400, code: 'BA001'}));
```
In conclusion, what I want to do is to send the server's error type as an object containing various information. Is there another way?

### Provide environment information

```bash
Binaries:
  Node: 20.11.0
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.1.4
  eslint-config-next: N/A
  react: 18
  react-dom: 18
  typescript: 5
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router

### Which stage(s) are affected? (Select all that apply)

next dev (local)

**Comment by coffeecupjapan** [comment]
@b0jun 

It seems like error message in server component is being protected by default in Next.js
see https://nextjs.org/docs/app/building-your-application/routing/error-handling#securing-sensitive-error-information

Or you can just return plain object in catch instead of returing error object I guess if you don't worry about sending any sensitive information in server to be shown at front-end.

**Comment by b0jun** [comment]
@coffeecupjapan

Thank you for your answer.

The biggest problem is that if I don't `throw error` in the catch of the server action, it won't be caught as an error in `Page Components`.
Should the task of screening for errors inside the phasor component be added?

---
If I try the following on the Api route, not Server Action, the desired error result value is normally output well.

```
import { NextResponse } from 'next/server';

const GET = async () => {
	try {
		throw new Error('ERROR', { cause: { status: 500 } });
	} catch (e: any) {
		return NextResponse.json({ message: e.message }, { status: e.cause.status });
	}
};

export { GET };
```

If I want to customize the error to the desired value and send it to the `Page Components`, is there no choice but to use `API Rotue`, Not `Server Action`?

**Comment by jjsenpai** [comment]
I found a workaround for this behaviour. Any error thrown from server to client side will contain generic message as mentioned in the docs : [Link](https://nextjs.org/docs/app/building-your-application/routing/error-handling#securing-sensitive-error-information)

What we can do instead is return the error in server side as json with status.

{
   //api
   const data = await response.json();
   if(!response.ok){
    
     return {status: response.status, ...data}
     
 }
 
 and in client side we can check if returnedObj.status === <code> and handle it as error.

**Comment by marcoshernanz** [comment]
Thanks for the report. I reproduced this on Next.js 14.1.4. React added support for transporting `Error.cause` in development in react/react#35810, and Next.js picked it up in #90211 for 16.2.0. The unchanged repro works on 16.3.2 and the current canary. Production redacts thrown server error details by design, so expected errors should be returned as serializable values. Closing as fixed in current releases.
